pkg/pmem-csi-driver/controllerserver-node.go

```diff
@@ -247,23 +262,20 @@ func (cs *nodeControllerServer) DeleteVolume(ctx context.Context, req *csi.Delet
 	defer nodeVolumeMutex.UnlockKey(req.VolumeId) //nolint: errcheck
 
 	klog.V(4).Infof("Node DeleteVolume: volumeID: %v", req.GetVolumeId())
-	eraseafter := true
-	if vol := cs.getVolumeByID(req.VolumeId); vol != nil {
-		// We recognize eraseafter=false/true, defaulting to true
-		if val, ok := vol.Params[pmemParameterKeyEraseAfter]; ok {
-			bVal, err := strconv.ParseBool(val)
-			if err != nil {
-				klog.Warningf("Ignoring invalid parameter %s:%s, reason: %s, falling back to default: %v",
-					pmemParameterKeyEraseAfter, val, err.Error(), eraseafter)
-			} else {
-				eraseafter = bVal
-			}
-		}
-	} else {
+	vol := cs.getVolumeByID(req.VolumeId)
+	if vol == nil {
+		// Already deleted.
 		return &csi.DeleteVolumeResponse{}, nil
 	}
+	parameters, err := parseVolumeParameters(nodeVolumeParameters, vol.Params)
```

Comment:
`vol.Params` are pre-validated and re-validating them here makes not much sense to me. Instead would prefer to change `Params` in `nodeVolume` struct from `map[stirng]string` to `volumeParameters`.

Reply:
`vol.Params` contains volume parameters that were potentially serialized by some older version of PMEM-CSI. We can't assume that they always match exactly with what currently is stored in `volumeParameters`.

If we allow backwards-incompatible changes, then this validation would be were we give up and return an error.

But as the next comment says, that would be very bad and force an admin to get involved. It'll be better to add support for old parameters into `parseVolumeParameters` such that they can still be converted into `volumeParameters` even if the parameter itself is deprecated and/or not allowed anymore for new volumes.

Reply:
In my opinion `parseVolumeParameters()` are supposed to call in two places:
 - while creating a (persisten/ephemeral/internal) volume - to validate params/context
 - while restoring volume info from state - to validate any backward-incompatibilites

and save the results in our cached voulume structures, Once after we should deal as Parameters.

Reply:
I started looking into that, but got stuck with the design of the statemanager: it deals with opaque objects that can be serialized to and from JSON. So `parseVolumeParameters` cannot be called there.

That means we have to keep `nodeVolume` as it is and validate at the level above statemanager. But then we have to be sure that we really catch all places where  `nodeVolume` is retrieved from the statemanager. The Go compiler will not warn us if we miss one.

I'd rather validate too often (i.e. before each use) instead of not often enough.

Reply:
Another advantage of validating before each use is that a failure is clearly associated with a certain volume and operation.

If we validate in `NewNodeControllerServer` (which is one place were JSON files are loaded), then all we can do is log some warning and ignore the existing volume. I'm almost certain that this would lead to leaked volume (as in `DeleteVolume` called, volume state couldn't be loaded and thus call succeeds, Kubernetes and the user think that the volume is gone whereas the actual state is "volume could not be deleted").

Reply:
> But then we have to be sure that we really catch all places where nodeVolume is retrieved from the statemanager.

... and then decide where to store the parsed parameters. Convert back to `map[string]string` and store them in the existing `nodeVolume` or create a `nodeInternalVolume` where `Params` is of type `volumeParameters` instead of `map[string]string`? Both is more complicated than what's currently in this PR.

Reply:
> That means we have to keep `nodeVolume` as it is and validate at the level above statemanager. But then we have to be sure that we really catch all places where `nodeVolume` is retrieved from the statemanager.

You can just provide json marshal interface methods to this new type `volumeParameters`:
```
func (p volumeParameters) MarshalJSON() ([]byte, error) {
    return json.Marshal(p.toVolumeContext())
}
```

This should be enough to save params using state manager.

Reply:
> ... and then decide where to store the parsed parameters. Convert back to `map[string]string` and store them in the existing `nodeVolume`

This is exactly what my proposal is. Deal everywhere as `volumeParameters` once after parsing.

Reply:
> You can just provide json marshal interface methods to this new type volumeParameters

That would work, except that it then has the problem I mentioned in https://github.com/intel/pmem-csi/pull/504#discussion_r363800540: suppose there is one "old" volume that for whatever reason cannot be handled anymore by the current PMEM-CSI version, which will be indicated by `parseVolumeParameters` by returning an error. Other volumes are fine.

In my current proposal, this will be noticed and reported in operations that deal with that volume. Those operations then fail in an obvious way and an admin can try to resolve the problem manually.

How do you propose to handle this aspect?

Reply:
If the driver is working with such an in-compatible state it would be good to fail as early as possible while node-driver initialization with an appropriate error message.

Reply:
But the *driver* is not in an incompatible state. Only some *volumes* are. It doesn't make sense to me to then also refuse to do any work for those volumes that are still okay.